test/e2e/framework/expect.go

```diff
@@ -40,6 +40,11 @@ func ExpectNoError(err error, explain ...interface{}) {
 	ExpectNoErrorWithOffset(1, err, explain...)
 }
 
+// ExpectNoErrorWithOffsetValue checks if "err" is set, and if so, fails assertion while logging the error at "offset" levels above its caller
+func ExpectNoErrorWithOffsetValue(err error, explain ...interface{}) {
```

Comment:
or ExpectNoErrorWithOffsetZero

Reply:
I am not particularly sold on the idea of writing a wrapper for `ExpectNoErrorWithOffset` in the context of theseTODOs being the best idea.

In terms of naming, `ExpectNoErrorWithOffsetZero` would be a more appropriate name compared to `ExpectNoErrorWithOffsetValue` because this would be a wrapper to `ExpectNoErrorWithOffset(0, err, explain...)`.
However, the name would be misleading as that 0 offset is converted to an offset of 1 within `ExpectNoErrorWithOffset`, ref: https://github.com/kubernetes/kubernetes/blob/master/test/e2e/framework/expect.go#L46.

All Expect* functions use a form of `gomega.ExpectWithOffset(1+offset)`, https://github.com/kubernetes/kubernetes/blob/master/test/e2e/framework/expect.go.
The motivation being that the Expect functions are designed with e2e tests in mind, so you want to surface the error assertion at the test level (error message shows up with the code that calls a function expecting the caller to be an e2e tests).

As an example of why the default offset of 1 in all the wrappers, lets consider this test
"Scheduler should continue assigning pods to nodes across restart" https://github.com/kubernetes/kubernetes/blob/master/test/e2e/apps/daemon_restart.go#L291.
It uses the default `framework.ExpectNoError()` wrappers which set the offset at 1.
In the case of a failure this is what we might see (https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-min-node-permissions/1233853232925642753)
```
[sig-apps] DaemonRestart [Disruptive] Scheduler should continue assigning pods to nodes across restart expand_less | 3m32s
-- | --
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/apps/daemon_restart.go:207
 Feb 29 21:32:25.833: Unexpected error:
     <*errors.errorString \| 0xc0028280f0>: {
         s: "1 pods disappeared for daemonrestart10-a4fdaa1f-4d82-4b42-ae7b-653392a903c0: daemonrestart10-a4fdaa1f-4d82-4b42-ae7b-653392a903c0-nrkxl",
     }
     1 pods disappeared for daemonrestart10-a4fdaa1f-4d82-4b42-ae7b-653392a903c0: daemonrestart10-a4fdaa1f-4d82-4b42-ae7b-653392a903c0-nrkxl
 occurred
 /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/apps/daemon_restart.go:222

```
This gives you the lines k8s.io/kubernetes/test/e2e/apps/daemon_restart.go:207 which point to a `ginkgo.BeforeEach(func() {...}`, https://github.com/kubernetes/kubernetes/blob/master/test/e2e/apps/daemon_restart.go#L207.
And it tells you the err occurred at k8s.io/kubernetes/test/e2e/apps/daemon_restart.go:222, https://github.com/kubernetes/kubernetes/blob/master/test/e2e/apps/daemon_restart.go#L222.
So it gives you the calleer and the place where you can figure out what Expect went wrong.

On the other hand, using an offset of 0, such as in the e2e test "Kubelet should not restart containers across restart" https://github.com/kubernetes/kubernetes/blob/master/test/e2e/apps/daemon_restart.go#L309, which uses `framework.ExpectNoErrorWithOffset(0, err)` may lead to an error of this sort (https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-ubuntu1-k8sstable1-serial/1232171621146431498):
```

[sig-apps] DaemonRestart [Disruptive] Kubelet should not restart containers across restart expand_less | 3m53s
-- | --
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:152
 Feb 25 06:03:29.433: All nodes should be ready after test, Not ready nodes: ", test-21b1dbcad3-minion-group-6bv5"
 /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:418
```
This tells you that something in the e2e test framework failed but it doesnt tell you who called it - failures make sense in the context of an e2e test.

---

With all that in mind, I think that calling an Expect* function with a 0 offset should be done at the writer's own risk - in most cases it wouldn't be good to call with with an offset of 0 because of the example listed above.
Having a wrapper for `ExpectWithOffset(0, err, explain)` may then make it easier for people to fall into a trap.
If for some reason people want that behaviour then they should be explicit about it.